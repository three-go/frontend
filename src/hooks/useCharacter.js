import React, { useState } from "react";

const useCharacter = (gameMap, score, setScore) => {
  const [position, setPosition] = useState({ x: 0, y: 0 });
  const [isValid, setIsValid] = useState(false);

  const moveLeft = () => {
    canMove("left") && setPosition({ ...position, x: position.x - 1 });
  };

  const moveRight = () => {
    canMove("right") && setPosition({ ...position, x: position.x + 1 });
  };

  const moveUp = () => {
    canMove("up") && setPosition({ ...position, y: position.y - 1 });
  };

  const moveDown = () => {
    canMove("down") && setPosition({ ...position, y: position.y + 1 });
  };

  const canMove = (direction) => {
    let isInBoundary = true;
    let nextPosition = -Infinity;
    let result = {};

    switch (direction) {
      case "left":
        isInBoundary = position.x - 1 > -1;
        nextPosition = gameMap[position.y][position.x - 1];
        break;

      case "right":
        isInBoundary = position.x + 1 < gameMap[0].length;
        nextPosition = gameMap[position.y][position.x + 1];
        break;

      case "up":
        isInBoundary = position.y - 1 > -1;
        nextPosition = gameMap[position.y - 1][position.x];
        break;

      case "down":
        isInBoundary = position.x - 1 > -1;
        nextPosition = gameMap[position.y][position.x - 1];
        break;
    }

    result = checkMove(isInBoundary, nextPosition);

    if (!result.canMove) {
      setScore(score - result.minusScore);
      setIsValid(false);
    } else {
      setIsValid(true);
    }

    return result.canMove;
  };

  const checkMove = (isInBoundary, nextPosition) => {
    if (isInBoundary && nextPosition > 0) {
      return { canMove: true, minusScore: 0 };
    }

    if (!isInBoundary) {
      return { canMove: false, minusScore: 20 };
    }

    if (!nextPosition) {
      return { canMove: false, minusScore: 10 };
    }
  };

  return {
    position,
    isValid,
    moveLeft,
    moveRight,
    moveDown,
    moveUp,
  };
};

export default useCharacter;
