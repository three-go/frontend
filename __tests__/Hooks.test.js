import React from "react";

import { fireEvent, render } from "@testing-library/react-native";
import { Button, View, Text } from "react-native";

import "@testing-library/jest-native";
import "react-native-gesture-handler/jestSetup";
import useCharacter from "../src/hooks/useCharacter";

const Test = () => {
  const character = useCharacter([
    [1, 0, 0],
    [1, 1, 1],
    [0, 0, 1],
    [0, 0, 2],
  ]);

  const handleMoveCharacter = () => {
    character.moveDown();
  };

  return (
    <View>
      <Button onPress={handleMoveCharacter} title="MOVE" testID="testButton" />
      );
      <Text testID="xPosition">{character.position.x}</Text>
      <Text testID="yPosition">{character.position.y}</Text>
      <Text testID="minusScore">{character.minusScore}</Text>
    </View>
  );
};

describe("Hooks Test", () => {
  it("useCharacter Test", () => {
    const test = render(<Test />);
    const testJson = test.toJSON();
    const button = test.getByTestId("testButton");
    const xPos = test.getByTestId("xPosition");
    const yPos = test.getByTestId("yPosition");
    const minusScore = test.getByTestId("minusScore");

    expect(button).toHaveTextContent("MOVE");
    expect(xPos).toHaveTextContent("0");
    expect(yPos).toHaveTextContent("0");
    expect(minusScore).toHaveTextContent("0");
    expect(testJson).toBeTruthy();
    expect(testJson).toMatchSnapshot();

    fireEvent.press(test.getByTestId("testButton"));

    expect(xPos).toHaveTextContent("0");
    expect(yPos).toHaveTextContent("1");
    expect(minusScore).toHaveTextContent("1");
  });
});
