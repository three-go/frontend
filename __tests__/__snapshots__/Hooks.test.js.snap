// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks Test useCharacter Test 1`] = `
<View>
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="testButton"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        MOVE
      </Text>
    </View>
  </View>
  );
  <Text
    testID="xPosition"
  >
    0
  </Text>
  <Text
    testID="yPosition"
  >
    0
  </Text>
  <Text
    testID="minusScore"
  >
    0
  </Text>
</View>
`;
